src/features/ci-cd/hooks/service-installs.ts

```diff
@@ -170,3 +171,33 @@ hooks.addPureHook('PATCH', 'resin', 'device', {
 		}
 	},
 });
+
+hooks.addPureHook('PATCH', 'resin', 'application', {
+	POSTRUN: async ({ api, request }) => {
+		const affectedIds = request.affectedIds!;
+		if (
+			request.values.should_be_running__release != null &&
+			affectedIds.length !== 0
+		) {
+			// Ensure that every device of the app we've just pinned, that is not itself pinned, has the necessary service install entries
+			const devices = (await api.get({
```

Comment:
Any reason you didn't just reuse part of the old implementation (doing a pine.get starting from the app) so that we don't do O(N) device GETs?

Reply:
Mostly because we have existing code to handle devices one-by-one and I wanted to reuse it - I think there are optimizations that can be made with bundling the queries from per-device into a single large batch which will cover both the app and device hooks but I think that makes sense to be a separate piece of work.

Also these two work by separate optimization methods, the app one fetched all the service installs of all existing devices and then diff'ed that vs the expected and added the missing ones, this code fetches all the services for a device where the service is missing and then creates just those, and I'm not sure which approach actually works out better (bearing in mind that there could be a huge number of pre-existing service installs that we're fetching, and I'd expect a much higher % pre-existing than missing in general)